crates/proto/src/lib.rs

```diff
@@ -48,3 +49,19 @@ pub use crate::xfer::retry_dns_handle::RetryDnsHandle;
 pub use crate::xfer::secure_dns_handle::SecureDnsHandle;
 #[doc(hidden)]
 pub use crate::xfer::{BufDnsStreamHandle, BufStreamHandle, MessageStreamHandle};
+
+use futures::Future;
+use tokio::runtime::Runtime;
+
+/// Generic executor.
+pub trait Executor {
+    /// Spawns a future object to run synchronously or asynchronously depending on the specific
+    /// executor.
+    fn run_until_completion<F: Future>(&mut self, future: F) -> F::Output;
```

Comment:
See below:
what is the future intention of the tests module (i.e. after this PR, what will be possible that is not today)?
-- I would like to reuse the test cases defined in tests module for fuchsia OS.
will itself eventually be marked as OS dependent? -- No The purpose is to make it OS independent.
will they be executor dependent? --No
how much code can be shared between the tests module and all the targets they will test? (as this is now it's hard for me to see how we'd generify the tests across different platforms)

Here is what I have in my private code base.
I created a lib inside Fuchsia, where fuchsia_async::net::UdpSocket and fuchsia_asyc::net::TcpStream implement traits defined in trust-dns-proto and fuchsia_async executor implements the Executor trait which is defined in this PR. When I define tests:
I am able to use the following codes
" #[test]
    fn test_udp_stream_ipv4() {
        use trust_dns_proto::tests::udp_stream_test;
        let exec = FuchsiaExec::new();
        udp_stream_test::<DnsUdpSocket, FuchsiaExec>(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), exec)
    }" for example.
So that this avoids the duplication of the test cases definition and make sure with fuchsia's specific types, the same test cases could pass.

For the Executor, please do look at #935
What would be your preference on landing these PR's? They seem to overlap a bit.
I am half-way of 935 PR, tried to understand what background was used for (done) and will try to understand the changes in the PR. 
What do you mean by preference, probably after I go through the #935 PR, I could have a better idea. Sorry for the delay.

Reply:
So basically, the fuschia related tests will live in  a separate batch of code, is that right? I think all of that makes sense, and so the rest of this change will work for that context.

The big concern I have is making sure I or others making changes don't break you, and that will be hard without some type of tests that enforces interfaces for you. We should definitely comment in the code interfaces that you need to use in Fuchsia so that we avoid the downstream breakage.

Maybe we could put a set of changes in the tests/integration crate that are clearly marked as testing the Fuchsia required APIs for compatibility?

Reply:
Fuchsia related tests code will live in the Fuchsia project. The reason is we prefer the trust-dns does not depend on Fuchsia crates at least for now. 
I will add comments for the interface definition.
The integration tests means trust-dns will use fuchsia's crates as dependency? If yes, probably we don't wanna that.

Reply:
> The integration tests means trust-dns will use fuchsia's crates as dependency? If yes, probably we don't wanna that.

No, I just meant that if we separate them out that they would be more apparent where we might break Fuchsia. Essentially, it would be an integration test that would use all the same test apis that we're exporting from the test module, just to guarantee better that they don't change. And people (me) might be less likely to accidentally change them if it's in an integration test suite...

Just an idea to try and future proof the integration there.

Reply:
I am updating the method name from run_until_completion to block_on. The reason I did not put as block_on was because in Fuchsia is all asynchronous. I guess I can put the right function call in the implementation itself anyway.

Reply:
Should that trait just be this one? and We add the spawn to this? 
I agree there are some overlap. It could live in one trait as behavior of the executor.

Reply:
Let's decide that in the other review... I think it's questionable if that trait is needed given your questions about ConnectionProvider potentially being enough...